interop-testing/src/test/java/io/grpc/testing/integration/CascadingTest.java

```diff
@@ -0,0 +1,334 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.testing.integration;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+import io.grpc.Context;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.ServerInterceptors;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.internal.ManagedChannelImpl;
+import io.grpc.internal.ServerImpl;
+import io.grpc.stub.StreamObserver;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+import java.io.IOException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Integration test for various forms of cancellation & deadline propagation.
+ */
+public class CascadingTest {
+
+  @Mock
+  TestServiceGrpc.TestService service;
+  private ManagedChannelImpl channel;
+  private ServerImpl server;
+  private AtomicInteger depth;
+  private int depthThreshold;
+  private AtomicInteger observedCancellations;
+  private AtomicInteger receivedCancellations;
+  private TestServiceGrpc.TestServiceBlockingStub blockingStub;
+  private TestServiceGrpc.TestServiceStub asyncStub;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    channel = InProcessChannelBuilder.forName("channel").build();
+    depthThreshold = 0;
+    depth = new AtomicInteger();
+    observedCancellations = new AtomicInteger();
+    receivedCancellations = new AtomicInteger();
+    blockingStub = TestServiceGrpc.newBlockingStub(channel);
+    asyncStub = TestServiceGrpc.newStub(channel);
+  }
+
+  @After
+  public void tearDown() {
+    Context.ROOT.attach();
+    channel.shutdownNow();
+    server.shutdownNow();
+  }
+
+  /**
+   * Test {@link Context} expiration propagates from the first node in the call chain all the way
+   * to the last.
+   */
+  @Test
+  public void testCascadingCancellationViaOuterContextExpiration() throws Exception {
+    startChainingServer();
+    Context.current().withDeadlineAfter(150, TimeUnit.MILLISECONDS).attach();
+    try {
+      blockingStub.unaryCall(Messages.SimpleRequest.getDefaultInstance());
+      fail("Expected cancellation");
+    } catch (StatusRuntimeException sre) {
+      // Wait for the workers to finish
+      Thread.sleep(100);
```

Comment:
Do you want to use directExecutor so you don't need to worry about timing caused by different threads?

If you want to use multiple threads you should provide your own executor, `shutdown()` it here, and then `awaitTermination()`.